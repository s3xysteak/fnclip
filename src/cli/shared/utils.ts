import type { PackageJson } from 'pkg-types'
import type { FnclipOptions } from './options'
import * as pkg from 'empathic/package'
import fs from 'fs-extra'
import * as path from 'pathe'
import { fnclipPath } from '../..'

export async function getMeta() {
  const res: Record<string, string> = await fs.readJson(path.join(fnclipPath, 'funcs-meta.json'))
  return res
}

export const exportContent = (f: string) => `export * from './${f}';\n`

export function ensureExt(fullname: string, ext: string) {
  return path.extname(fullname) ? fullname : `${fullname}${ext}`
}

export async function isTypescript(cwd: string) {
  const packageJsonPath = pkg.up({ cwd })
  if (!packageJsonPath)
    return

  const obj: PackageJson = await fs.readJson(packageJsonPath)
  return !!(obj.dependencies?.typescript || obj.devDependencies?.typescript)
}

export async function updateIndex(opts: FnclipOptions) {
  if (!opts.index)
    return

  const dirPath = path.join(opts.cwd, opts.dir)
  const indexPathMaybeExt = path.join(dirPath, opts.indexPath)

  // check exist index file
  let indexRealPath: string
  for (const ext of ['.ts', '.js']) {
    const targetPath = ensureExt(indexPathMaybeExt, ext)
    if (await fs.exists(targetPath)) {
      indexRealPath = targetPath
      break
    }
  }

  indexRealPath ??= ensureExt(indexPathMaybeExt, (opts.ts ? '.ts' : '.js'))
  await fs.ensureFile(indexRealPath)

  const funcs = await fs.readdir(dirPath)
  await fs.writeFile(indexRealPath, addIgnoreToContent(funcs.map(exportContent).join('')))
}

export function addIgnoreToContent(content: string) {
  const comments = `/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck

`
  return content.includes(comments) ? content : comments + content
}
