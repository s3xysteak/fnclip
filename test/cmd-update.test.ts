import fs from 'fs-extra'
import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import { rootJoin, run } from './shared'

describe('update', () => {
  const BASE = 'test-update'
  const join = rootJoin.createJoin(BASE)

  beforeAll(async () => {
    await run(`pnpm fnclip add objectKeys isInRange --dir ${BASE}`)
    const files = await fs.readdir(join())
    await Promise.all(files.map(async (pathname) => {
      await fs.writeFile(join(pathname), `// ${pathname}`)
    }))
  })
  afterAll(async () => {
    await fs.rm(join(), { recursive: true, force: true })
  })

  it('should work', async () => {
    const funcs = await fs.readdir(join())
    const map = (await Promise.all(
      funcs.map(async pathname => ({
        pathname,
        content: await fs.readFile(join(pathname), 'utf8'),
      })),
    )).sort((a, b) => a.pathname.localeCompare(b.pathname))

    await run(`pnpm fnclip update --dir ${join()}`)

    for (const { pathname } of map) {
      const content = await fs.readFile(join(pathname), 'utf8')
      expect(content).toMatchSnapshot()
    }

    await Promise.all(map.map(async ({ pathname, content: rawContent }) => {
      await fs.writeFile(join(pathname), rawContent)
    }))
  })
})
