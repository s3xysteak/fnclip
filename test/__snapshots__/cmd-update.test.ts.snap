// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`update > should work 1`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

export * from './isInRange';
export * from './objectKeys';"
`;

exports[`update > should work 2`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

//#region src/functions/number/isInRange/index.d.ts
/**
 * Is a number in a range
 *
 * @example
 *
 * \`\`\`js
 * isInRange(5, [1, 10]) // -> true
 * \`\`\`
 *
 * @see {@link https://s3xysteak.github.io/fnclip/functions/number/isInRange/}
 */
declare function isInRange(val: number, [a, b]: [number, number]): boolean;
//#endregion
export { isInRange };"
`;

exports[`update > should work 3`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

//#region src/functions/number/isInRange/index.ts
/**
* Is a number in a range
*
* @example
*
* \`\`\`js
* isInRange(5, [1, 10]) // -> true
* \`\`\`
*
* @see {@link https://s3xysteak.github.io/fnclip/functions/number/isInRange/}
*/
function isInRange(val, [a, b]) {
	return val >= Math.min(a, b) && val <= Math.max(a, b);
}

//#endregion
export { isInRange };"
`;

exports[`update > should work 4`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

//#region src/functions/object/objectKeys/index.d.ts
/**
 * Strict typed \`Object.keys\`
 *
 * @link https://github.com/antfu/utils/blob/main/src/object.ts
 * @see {@link https://s3xysteak.github.io/fnclip/functions/object/objectKeys/}
 */
declare function objectKeys<T extends object>(obj: T): Array<\`\${keyof T & (string | number | boolean | null | undefined)}\`>;
//#endregion
export { objectKeys };"
`;

exports[`update > should work 5`] = `
"/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// biome-ignore lint: disable
// Generated by fnclip

//#region src/functions/object/objectKeys/index.ts
/**
* Strict typed \`Object.keys\`
*
* @link https://github.com/antfu/utils/blob/main/src/object.ts
* @see {@link https://s3xysteak.github.io/fnclip/functions/object/objectKeys/}
*/
function objectKeys(obj) {
	return Object.keys(obj);
}

//#endregion
export { objectKeys };"
`;
